editors/code/src/test/format.test.ts

```diff
@@ -0,0 +1,67 @@
+import * as vscode from "vscode";
+import * as assert from "assert";
+import * as fs from "fs";
+import { before } from "mocha";
+
+import { withFileEditor } from "./editor-utils";
+import { snapshotPath, testPath, waitLsp, withToml } from "./extension";
+
+before(async () => {
+	await waitLsp();
+
+	// Check for an `air.toml` file that is not properly cleaned up as it would
+	// interfere with tests
+	const airTomlPath = testPath("air.toml");
+	if (fs.existsSync(airTomlPath)) {
+		console.log(
+			"Deleting `air.toml` file that was not properly cleaned up",
+		);
+		fs.unlinkSync(airTomlPath);
+	}
+});
+
+suite("Format Test Suite", () => {
+	test("Format document with forward propagation of settings", async () => {
+		await withFileEditor(snapshotPath("format.R"), async (editor, doc) => {
+			const old = doc.getText();
+
+			assert.strictEqual(editor.options.insertSpaces, true);
+			assert.strictEqual(editor.options.indentSize, 8);
+			assert.strictEqual(editor.options.tabSize, 8);
+
+			await vscode.commands.executeCommand(
+				"editor.action.formatDocument",
+			);
+			await doc.save();
+
+			assert.strictEqual(doc.getText(), old);
```

Comment:
I'm not sure I feel like this test is strong enough. It doesn't test that anything actually happened. i.e. 

```
await vscode.commands.executeCommand(
  "editor.action.formatDocument",
);
```

this could be a no-op and the test would pass, which makes me uncomfortable. I feel like we need two sets of files?
- One folder of fixtures, i.e. `fixtures/format.R`, and we format these and then read their contents
- One folder of snapshots, i.e. `snapshots/format.R` and we compare the formatted contents above to these contents

I don't think we need to auto generate the snapshot files like testthat does, we can manually make the before/after versions of `format.R` ourselves, but I do feel it is important to test that we are actually getting invoked, forcing _something_ to happen.

Reply:
Good point. I'm now stripping whitespace from the input files. If the formatter is a no-op and didn't repair the missing whitespace the tests will now fail.